File: src/data/fetcher.py
"""
Sector Data Fetcher
Fetches price and volume data for sector ETFs using yfinance
"""

import yfinance as yf
import pandas as pd
from datetime import datetime, timedelta

SECTOR_ETFS = ['XLK', 'XLF', 'XLV', 'XLE', 'XLI', 'XLP', 'XLU', 'XLRE', 'XLC', 'XLB', 'SPY']

def fetch_sector_data(start_date=None, end_date=None):
    """Fetch price and volume data for all sector ETFs"""
    if start_date is None:
        start_date = (datetime.now() - timedelta(days=365)).strftime('%Y-%m-%d')
    if end_date is None:
        end_date = datetime.now().strftime('%Y-%m-%d')
    
    print(f"Fetching sector data from {start_date} to {end_date}...")
    
    data = {}
    for symbol in SECTOR_ETFS:
        try:
            ticker = yf.Ticker(symbol)
            df = ticker.history(start=start_date, end=end_date)
            if not df.empty:
                data[symbol] = df
                print(f"  {symbol}: {len(df)} rows")
        except Exception as e:
            print(f"  {symbol}: Error - {e}")
    
    return data

def calculate_money_flow(prices, volumes):
    """Calculate money flow = price * volume"""
    return prices * volumes

def calculate_sector_flows(data):
    """Calculate sector-to-sector money flow"""
    flows = {}
    for symbol, df in data.items():
        df['money_flow'] = df['Close'] * df['Volume']
        df['flow_change'] = df['money_flow'].diff()
        flows[symbol] = df[['Close', 'Volume', 'money_flow', 'flow_change']]
    return flows

if __name__ == '__main__':
    data = fetch_sector_data()
    flows = calculate_sector_flows(data)
    print(f"\nProcessed {len(flows)} sectors")

File: src/data/fetcher.js
/**
 * Sector Data Fetcher
 * Fetches ETF price and volume data using yfinance
 */

import yfinance as yf
import pandas as pd
from datetime import datetime, timedelta

SECTOR_ETFS = ['XLK', 'XLF', 'XLV', 'XLE', 'XLI', 'XLP', 'XLU', 'XLRE', 'XLC', 'XLB', 'SPY']

def fetch_sector_data(period='1y', interval='1d'):
    """Fetch price and volume data for all sector ETFs"""
    data = {}
    
    for symbol in SECTOR_ETFS:
        print(f"Fetching {symbol}...")
        ticker = yf.Ticker(symbol)
        df = ticker.history(period=period, interval=interval)
        
        if df.empty:
            print(f"  No data for {symbol}")
            continue
        
        df = df.reset_index()
        df.columns = [c.lower().replace(' ', '_') for c in df.columns]
        data[symbol] = df
        print(f"  Got {len(df)} rows")
    
    return data

def calculate_money_flow(data):
    """Calculate money flow for each sector"""
    flows = {}
    
    for symbol, df in data.items():
        df = df.copy()
        # Money Flow = Close × Volume
        df['money_flow'] = df['close'] * df['volume']
        # Flow change
        df['flow_change'] = df['money_flow'].diff()
        df['flow_change_pct'] = df['flow_change'].pct_change() * 100
        flows[symbol] = df
    
    return flows

def aggregate_by_period(data, period='week'):
    """Aggregate data by time period"""
    aggregated = {}
    
    for symbol, df in data.items():
        df = df.copy()
        df['date'] = pd.to_datetime(df['date'] if 'date' in df.columns else df.index)
        df = df.set_index('date')
        
        if period == 'day':
            agg = df
        elif period == 'week':
            agg = df.resample('W').agg({
                'open': 'first',
                'high': 'max',
                'low': 'min',
                'close': 'last',
                'volume': 'sum',
                'money_flow': 'sum'
            })
        elif period == 'month':
            agg = df.resample('M').agg({
                'open': 'first',
                'high': 'max',
                'low': 'min',
                'close': 'last',
                'volume': 'sum',
                'money_flow': 'sum'
            })
        elif period == 'quarter':
            agg = df.resample('Q').agg({
                'open': 'first',
                'high': 'max',
                'low': 'min',
                'close': 'last',
                'volume': 'sum',
                'money_flow': 'sum'
            })
        elif period == 'year':
            agg = df.resample('Y').agg({
                'open': 'first',
                'high': 'max',
                'low': 'min',
                'close': 'last',
                'volume': 'sum',
                'money_flow': 'sum'
            })
        
        agg['flow_change'] = agg['money_flow'].diff()
        aggregated[symbol] = agg
    
    return aggregated

if __name__ == "__main__":
    print("Fetching sector data...")
    data = fetch_sector_data(period='2y', interval='1d')
    
    print("\nCalculating money flow...")
    flows = calculate_money_flow(data)
    
    print("\nAggregating by week...")
    weekly = aggregate_by_period(flows, 'week')
    
    print("\nSample output for XLK:")
    print(weekly['XLK'].tail())
